### PR TITLE
Minecraft: recipe update & other enhancements

### LINKED ISSUE #1988: Minecraft: should support alternate names for items.
Item names are too hard to guess. They should have alternate names or should be stored with the most common term. 'shovel' for example, isn't in the item list, but rather stored in the plural form as 'shovels'. Some people may say "how to craft shovels" whereas others might prefer "how to craft a shovel". I doubt anyone will guess "Jack-O-Lantern".

@engvik 

------
IA Page:  http://duck.co/ia/view/minecraft

### PR DESCRIPTION
**Pull request by 122131**
###### What does your Pull Request do (check all that apply)?

Choose the most relevant items and use the following title template to name
your Pull Request.

- [ ] New Instant Answer
    - [ ] Cheat Sheets: **`New {Cheat Sheet Name} Cheat Sheet`**
    - [ ] Other: **`New {IA Name} Instant Answer`**
- [X] Improvement
    - [ ] Bug fix: **`{IA Name}: Fix {Issue number or one-line description}`**
    - [X] Enhancement: **`{Minecraft}: {updated recipes, various other fixes and improvements}`**
- [ ] Non–Instant Answer
    - [ ] Other (Role, Template, Test, Documentation, etc.): **`{GoodieRole/Templates/Tests/Docs}: {Short Description}`**

###### Description of changes

- Updated recipes, blocks and items to Minecraft 1.10
- Added alternate names for recipes
- Updated some recipe images with local GIFs

###### Which issues (if any) does this fix?

Fixes #1988 

###### People to notify @ecounysis 

---

Instant Answer Page: https://duck.co/ia/view/minecraft

[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @engvik

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by 122131** [comment]
Still WIP.
To do:
- [ ] finish adding new recipes and changing the old ones 
- [ ] finish adding regex values

**Comment by edgesince84** [comment]
thanks @122131 - now I want some cake!

**Comment by 122131** [comment]
@edgesince84 Me too, but before I have to finish with this! 😄

**Comment by engvik** [comment]
Nice work, @122131 :smile:

Just ping me when you're done! FYI, I will be on vacation and unavailable from the 28th of June until July 5th.

**Comment by 122131** [comment]
@engvik I'll try to complete this by that day 👍 
P.S. happy holidays! 😄

File: share/goodie/minecraft/crafting-guide.json [hunk 1]
```diff
@@ -2,945 +2,1171 @@
...
+      "ingredients": "1 Gold Ingot",
       "description": "Used to craft Glistering Melon. Dropped by Zombie Pigman. Can be crafted into Gold Ingots."
     },
     {
       "name": "Magma Cream",
+      "regex": "(?:magma ?creams?)",
       "image": "http://www.minecraftxl.com/images/wiki/recipes/magma-cream-crafting.png",
-      "ingredients": "Slimeball + Blaze Powder",
+      "ingredients": "1 Slimeball + 1 Blaze Powder",
       "description": "Used in Potions, mostly
```
Comment (moollaza) [inline share/goodie/minecraft/crafting-guide.json]: @122131 looks like you have a double quote here, causing problems!

Comment (moollaza) [inline share/goodie/minecraft/crafting-guide.json]: Once that's fixed the tests should pass again and I can deploy this to beta 👍

**Comment by 122131** [comment]
Some block images are still unavailable because I have to create them manually.
I haven't seen any png/gif image of minecraft 1.9/1.10 recipes online, so if someone has a link feel free to share it! 👍

**Comment by 122131** [comment]
@moollaza  I don't know why after 8730ecb478d3cd066dc1ad10b418a88c7c5af328 some words like "chest" or "bread" don't trigger the IA anymore, while expressions like "mushroom soup" trigger correctly. 😕  Any help?

**Comment by moollaza** [comment]
@122131 it looks like your JSON is broken ;) -- there's a double `""`

**Comment by 122131** [comment]
@moollaza Where? I can't see it 😕

**Comment by moollaza** [comment]
@122131 my mistake, looks like it's been fixed. I was looking at an old commit

**Comment by moollaza** [comment]
@122131 one or more of your regex are broken. If you log the matched regex, it keeps showing the window panes regex because it contains `(|` along with `(glass?)` meaning it will capture a space in a query.

You need to repair that one and then keep trying to see what's happening 👍

**Comment by 122131** [comment]
@moollaza ok, I will check all the regex!
Thanks for the help! 😄

File: lib/DDG/Goodie/Minecraft.pm [hunk 2]
```diff
@@ -20,17 +20,25 @@ my %recipes = map{ lc $_->{'name'} => $_ } (@{ $decoded->{'items'} });
...
     my $recipe;
     # find recipe name
     foreach my $recipe_name (@recipe_names) {
-        if ($remainder =~ s/$recipe_name//i) {
-            $recipe = $recipes{$recipe_name};
-            last;
+        my $regex = $recipe_name; 
+        
+        # check if we have a regex for the recipe 
+        if (exists( $recipes{$recipe_name}->{'regex'}))
```
Comment (moollaza) [inline lib/DDG/Goodie/Minecraft.pm]: nit: we typically keep the brace on the same line. The rest of the code here does this so we should follow suite.

You could simplify this to:

```perl
my $regex = $recipes{$recipe_name}->{'regex'} // $recipe_name;
```

File: share/goodie/minecraft/crafting-guide.json [hunk 3]
```diff
@@ -2,945 +2,1192 @@
...
     {
       "name": "Magma Cream",
+      "regex": "(?:magma ?creams?)",
       "image": "http://www.minecraftxl.com/images/wiki/recipes/magma-cream-crafting.png",
-      "ingredients": "Slimeball + Blaze Powder",
+      "ingredients": "1 Slimeball + 1 Blaze Powder",
       "description": "Used in Potions, mostly for fire resistance."
     },
     {
       "name": "Bone Meal",
```
Comment (moollaza) [inline share/goodie/minecraft/crafting-guide.json]: Do we plan to have images for these? If not we should leave the value blank and not define the `image` property, so we don't create an `img` tag on the page

File: lib/DDG/Goodie/Minecraft.pm [hunk 4]
```diff
@@ -50,11 +58,11 @@ handle remainder => sub {
             title => $recipe->{'name'},
             subtitle => "Ingredients: " . $recipe->{'ingredients'},
             description => $recipe->{'description'},
-            image => 'https://duckduckgo.com/iu/?u=' . uri_esc( $recipe->{'image'} )
+            image => $recipe->{'image'}
```
Comment (moollaza) [inline lib/DDG/Goodie/Minecraft.pm]: Can you please try adding `imageTile => 1` -- this should force the images to not be clipped!

https://github.com/duckduckgo/duckduckgo-answerbar-templates/blob/master/detail/basic_info_detail.handlebars#L5

**Comment by moollaza** [comment]
@122131 this is looking really good! 💯 

There's a couple minor fixes to make and then I'd be happy to merge this 👍

File: share/goodie/minecraft/crafting-guide.json (hunk 3 above)
Comment (122131) [inline share/goodie/minecraft/crafting-guide.json]: I'm trying to add an image for every single item/block in the list, but I have to create them manually, unless someone has a link with the images already done!

File: lib/DDG/Goodie/Minecraft.pm (hunk 4 above)
Comment (122131) [inline lib/DDG/Goodie/Minecraft.pm]: Is this a try to fix #1124 ?

**Comment by 122131** [comment]
@engvik @moollaza  I think it will take some other days for me to complete the images and considered that I will be on holiday next week, I'll try to finish everything by the middle of July. 😄 
Sorry in advance for the delay!

**Comment by engvik** [comment]
No worries :smile: Thanks again for the great contribution!

**Comment by moollaza** [comment]
> Sorry in advance for the delay!

@122131 no worries! Please enjoy your vacation!

**Comment by 122131** [comment]
@moollaza It seems that the trigger words "what is" don't work (the IA doesn't trigger when I search for "what is the recipe for cake in minecraft", but when I search for "how do I craft a cake in minecraft" it's OK.
Any suggestion? 😕

**Comment by engvik** [comment]
I had a quick look, and I can't immediately see what wrong :/

**Comment by moollaza** [comment]
@122131 you need to add `\b` around `if ($remainder =~ s/\b$regex\b//i) {` -- without that you're matching `hat` in the word `what` and it's causing problems 👍 

With `\b` it works as expected

File: lib/DDG/Goodie/Minecraft.pm [hunk 5]
```diff
@@ -20,18 +20,20 @@ my %recipes = map{ lc $_->{'name'} => $_ } (@{ $decoded->{'items'} });
...
-    # find recipe name
+    # find recipe name and regex
     foreach my $recipe_name (@recipe_names) {
-        if ($remainder =~ s/$recipe_name//i) {
-            $recipe = $recipes{$recipe_name};
-            last;
-        }
+        my $regex = $recipes{$recipe_name}->{'regex'} // $recipe_name;
+
+        if ($remainder =~ s/$regex//i) {
```
Comment (moollaza) [inline lib/DDG/Goodie/Minecraft.pm]: Should be `if ($remainder =~ s/\b$regex\b//i) { `

**Comment by 122131** [comment]
@moollaza Thanks! 😄

**Comment by 122131** [comment]
I was thinking on how we could keep this updated... The only way I know (as suggested by @moollaza in https://github.com/duckduckgo/zeroclickinfo-fathead/issues/213) is to have a GamePedia data dump and create a Fathead IA, but I don't think it is so straightforward... Any thoughts?

**Comment by 122131** [comment]
Ready for review and merging :smiley:

**Comment by engvik** [comment]
Great work, @122131! This LGTM :+1:

**Comment by moollaza** [comment]
@122131 this is **_JUST_** about done but there's 1 last issue -- the images aren't going to load because we're using an unversioned path. For the images to work in Production the path needs to contain a version number.

You can get the current version in the Perl like so:

https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/SolarSystem.pm#L14

Then you'll want to build an image path that looks like: `"/share/goodie/minecraft/$version/images/chest.gif"

I would suggest that you change `image` to `imageName` or similar and build your path in the Perl

**Comment by moollaza** [comment]
Here's an example of a broken image on Beta: https://beta.duckduckgo.com/?q=minecraft+coarse+dirt+recipe&t=ha&ia=answer

**Comment by 122131** [comment]
@moollaza In the Travis Build I see some Warnings like
Use of uninitialized value...

How can I fix it? :confused:

File: lib/DDG/Goodie/Minecraft.pm [hunk 6]
```diff
@@ -8,40 +8,54 @@ use JSON::MaybeXS;
...
     }
-    return if $remainder && $remainder !~ m/^\s+$/; # return if we had leftover unwanted words (not just whitespace)
+    return if $remainder && $remainder !~ m/^\s+$/; # Return if we had leftover unwanted words (not just whitespace)
+    
+    # Get image correct path
+    my $image;
+    my $imageName = $recipe->{'imageName'};
+    my $localImage = $recipe->{'localImage'};
+    
+    if ($localImage =~ "yes") { # If local image is used
```
Comment (moollaza) [inline lib/DDG/Goodie/Minecraft.pm]: `if ($localImage eq "yes")`

File: lib/DDG/Goodie/Minecraft.pm [hunk 7]
```diff
@@ -8,40 +8,54 @@ use JSON::MaybeXS;
...
     foreach my $extra_word (@extra_words) {
         $remainder =~ s/\b$extra_word\b//gi;
     }
-    return if $remainder && $remainder !~ m/^\s+$/; # return if we had leftover unwanted words (not just whitespace)
+    return if $remainder && $remainder !~ m/^\s+$/; # Return if we had leftover unwanted words (not just whitespace)
+    
+    # Get image correct path
+    my $image;
+    my $imageName = $recipe->{'imageName'};
+    my $localImage = $recipe->{'localImage'};
```
Comment (moollaza) [inline lib/DDG/Goodie/Minecraft.pm]: `my $localImage = $recipe->{'localImage'} // "no"`

**Comment by moollaza** [comment]
> How can I fix it? 😕

@122131 `$localImage` is only defined when that key exists in `$recipe` so `my $localImage = $recipe->{'localImage'};` doesn't always return a value.

This means `$localImage` can be `undef` when you check it's value later, so you need to give it a default value to ensure it's always defined.

File: lib/DDG/Goodie/Minecraft.pm (hunk 7 above)
Comment (moollaza) [inline lib/DDG/Goodie/Minecraft.pm]: @122131 This will solve the `uninitialized` warnings 👍

**Comment by 122131** [comment]
@moollaza Thanks (as always :laughing:) for the help, I'll make the changes immediately! :+1:

**Comment by moollaza** [comment]
@122131 Thanks! This LGTM :+1:

**Comment by jomo** [comment]
Could the wiki and image links be switched to https instead of http?

**Comment by 122131** [comment]
@jomo Last time I checked, a premium account was needed for HTTPS connections to Gamepedia subdomains.
For the images, HTTPS is available, I can make a new pull request :+1:

**Comment by jomo** [comment]
> a premium account was needed for HTTPS connections

What the actual …

When I'm logged in, https://minecraft.gamepedia.com/Block_of_Redstone works fine (apparently I have a premium account? 😮). When I'm logged out, it just redirects to HTTPS, but still works. I think linking to HTTPS is preferable in any case.

**Comment by 122131** [comment]
@jomo Well, I don't have a Gamepedia account, so I found [this page](https://www.gamepedia.com/PRO) and I thought a PRO account was needed. :confused: 

I will anyhow change the source to HTTPS, since it still works. Thanks for the suggestion! :smiley:

**Pull request merged**